File: 6/count_binary_substr.py
class Solution:
    def countBinarySubstrings(self, s: str) -> int:
        m = map(len, s.replace('01', '0 1').replace('10', '1 0').split())
        l = list(m)
        return sum(min(a, b) for a, b in zip(l, l[1:]))


s = Solution()
print(s.countBinarySubstrings("00110"))
print(s.countBinarySubstrings("00110011"))

File: 6/count_binary_substr.go
package main

import (
	"fmt"
)

func countBinarySubstrings(s string) int {
	cur := 1
	pre := 0
	res := 0
	for i := 1; i < len(s); i++ {
		if s[i-1] == s[i] {
			cur++
		} else {
			res += min(cur, pre)
			pre = cur
			cur = 1
		}
	}
	return res + min(cur, pre)
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	fmt.Println(countBinarySubstrings("00110"))
	fmt.Println(countBinarySubstrings("00110011"))
}
